core/transaction-pool/src/lib.rs

```diff
@@ -20,10 +20,246 @@
 #![warn(unused_extern_crates)]
 
 mod api;
+mod maintainer;
 #[cfg(test)]
 mod tests;
 
 pub mod error;
 
-pub use api::{FullChainApi, LightChainApi};
 pub use txpool;
+pub use crate::api::{FullChainApi, LightChainApi};
+pub use crate::maintainer::{
+	TransactionPoolMaintainer,
+	DefaultFullTransactionPoolMaintainer, DefaultLightTransactionPoolMaintainer,
+};
+
+use std::{collections::HashMap, hash::Hash, sync::Arc};
+use futures::Future;
+
+use client::{
+	Client,
+	light::fetcher::Fetcher,
+	runtime_api::TaggedTransactionQueue,
+};
+use primitives::{Blake2Hasher, H256};
+use sr_primitives::{
+	serde::Serialize,
+	generic::BlockId,
+	traits::{Block as BlockT, Member, ProvideRuntimeApi},
+};
+
+use txpool::{BestIterator, EventStream, Options, watcher::Watcher};
+
+/// Extrinsic hash type for a pool.
+pub type ExHash<P> = <P as TransactionPool>::Hash;
+/// Block hash type for a pool.
+pub type BlockHash<P> = <<P as TransactionPool>::Block as BlockT>::Hash;
+/// Extrinsic type for a pool.
+pub type ExtrinsicFor<P> = <<P as TransactionPool>::Block as BlockT>::Extrinsic;
+
+/// Transaction pool interface.
+pub trait TransactionPool: Send + Sync {
+	/// Block type.
+	type Block: BlockT;
+	/// Transaction Hash type.
+	type Hash: Hash + Eq + Member + Serialize;
+	/// Error type.
+	type Error: From<txpool::error::Error> + txpool::error::IntoPoolError;
+
+	/// Returns a future that imports a bunch of unverified extrinsics to the pool.
+	fn submit_at(
+		&self,
+		at: &BlockId<Self::Block>,
+		xts: impl IntoIterator<Item=ExtrinsicFor<Self>> + 'static,
+		force: bool,
+	) -> Box<dyn Future<Output=Result<Vec<Result<Self::Hash, Self::Error>>, Self::Error>> + Send + Unpin>;
+
+	/// Returns a future that imports one unverified extrinsic to the pool.
+	fn submit_one(
+		&self,
+		at: &BlockId<Self::Block>,
+		xt: ExtrinsicFor<Self>,
+	) -> Box<dyn Future<Output=Result<Self::Hash, Self::Error>> + Send + Unpin>;
+
+	/// Returns a future that import a single extrinsic and starts to watch their progress in the pool.
+	fn submit_and_watch(
+		&self,
+		at: &BlockId<Self::Block>,
+		xt: ExtrinsicFor<Self>,
+	) -> Box<dyn Future<Output=Result<Watcher<Self::Hash, BlockHash<Self>>, Self::Error>> + Send + Unpin>;
+
+	/// Remove extrinsics identified by given hashes (and dependent extrinsics) from the pool.
+	fn remove_invalid(
+		&self,
+		hashes: &[Self::Hash],
+	) -> Vec<Arc<txpool::base_pool::Transaction<Self::Hash, ExtrinsicFor<Self>>>>;
+
+	/// Returns pool status.
+	fn status(&self) -> txpool::base_pool::Status;
+
+	/// Get an iterator for ready transactions ordered by priority
+	fn ready(&self) -> BestIterator<Self::Hash, ExtrinsicFor<Self>>;
```

Comment:
Hmm, would be best to avoid leaking `BestIterator` from the crate. Maybe let's just `Box` it here as well.